### ISSUE TITLE
bug: `no-raw-colors` reports custom `--text-*` font sizes as undeclared theme colors

### ISSUE THREAD

**Issue by salmundani** [shadcn-ui/lint#8]
The following report is AI-generated, let me know if you need more details!

## Summary

When a Tailwind v4 theme defines a custom font-size scale with `--text-*`
variables, `no-raw-colors` reports those `text-<name>` classes as undeclared
theme colors. Tailwind reads them as font sizes, but the linter reads them as
text colors.

## Reproduction

`src/index.css` (the `tailwind.css` entry in `components.json`):

```css
@import "tailwindcss";

@theme inline {
  --color-foreground: var(--foreground);
}

@theme {
  --text-heading: 1.875rem;
  --text-heading--line-height: 0.9;
  --text-body: 0.875rem;
  --text-body--line-height: 1.4;
}
```

`src/page.tsx`:

```tsx
export function Page() {
  return (
    <>
      <h1 className="text-heading">Title</h1>
      <p className="text-body text-foreground">Body</p>
    </>
  )
}
```

`eslint.config.js`:

```js
rules: {
  "shadcn/no-raw-colors": "error",
}
```

## Expected

No reports. Tailwind generates `text-heading` and `text-body` as font-size
utilities from `--text-heading` and `--text-body`.

## Actual

```text
"text-heading" is not a declared theme color. Use one of: foreground (+…). To add a color, declare --color-<name> in src/index.css first
"text-body" is not a declared theme color. Use one of: foreground (+…). To add a color, declare --color-<name> in src/index.css first
```

## Cause

The class group comes from `cn`'s `defaultConfig()`, not from the theme. That
config only knows Tailwind's built-in font sizes (`text-xs` to `text-9xl`), so
any other `text-<word>` falls into the `text-color` group. The `judge` step in
`no-raw-colors` then sees the color category, finds `body` missing from the
declared `--color-*` tokens, and reports it.

`isTypoOfAnotherUtility` doesn't catch it either: Tailwind generates
`text-body`, so there is no suggestion and the class is still reported.

The theme's `--text-*` variables are read for the arbitrary-value scale
suggestions, but they don't affect how a class is classified.

A project's own `cn` doesn't help. The linter calls `defaultConfig()` with no
arguments, so a project extension, such as tailwind-merge's
`extendTailwindMerge({ extend: { classGroups: { "font-size": [{ text: ["body"] }] } } })`,
never reaches the linter.

`no-restyle` uses the same classification, so it should also count
`text-heading` on a component as a color override instead of typography.
A contract with `allow: ["layout", "typography"]` would then not cover it.

## Possible fixes

- Classify `text-<name>` as `font-size` when the theme declares
  `--text-<name>` and not `--color-<name>`.
- Or let settings add to the class groups, for example
  `settings.shadcn.classGroups`, and merge those into the `cn` config.

## Workaround

List the classes as exceptions:

```js
"shadcn/no-raw-colors": ["error", {
  allow: ["text-heading", "text-body"],
}]
```

## Environment

- `@shadcn/lint` 0.1.0
- `cn` 0.2.6 (bundled; the project uses `clsx` + `tailwind-merge`)
- ESLint 9.39.5
- Tailwind CSS 4.3.3
- Node.js 24.16.0

**Comment by camdressler** [comment]
Same cause seen from the other theme namespaces, plus the `no-restyle` side of it. Posting here rather than as a new issue since #8 already has the `--text-*` half.

### What happens

Classification comes from `cn`'s `defaultConfig()`, which is called with no arguments, so it only knows Tailwind's built-in scale names. Theme keys outside `--color-*` never reach it. A `text-<name>` or `shadow-<name>` the theme declares falls into a colour group and is reported as a raw colour, and a `px-<name>` or `rounded-<name>` has no category at all, so `no-restyle` reports it whatever is in `allow`.

### Minimal reproduction

`src/app.css`

```css
@import "tailwindcss";

@theme {
  --text-caption: 0.75rem;
  --shadow-panel: 0 1px 2px rgb(0 0 0 / 0.1);
  --spacing-gutter: 12px;
  --radius-tight: 2px;
}
```

`src/page.tsx`, where `Card` is a local design-system component:

```tsx
<Card className="text-caption shadow-panel px-gutter rounded-tight">Caption</Card>
```

with `"shadcn/no-raw-colors": "warn"` and `"shadcn/no-restyle": ["warn", { "allow": ["layout", "spacing"] }]`.

### Actual

- `no-raw-colors` reports `text-caption` and `shadow-panel` as undeclared theme colours and suggests declaring `--color-caption` and `--color-panel`, which would be the wrong declaration.
- `no-restyle` reports `px-gutter` and `rounded-tight` as unclassified even though `spacing` is allowed. The only way through is a wildcard entry such as `px-*` in `allow`, which gives up the check for that whole utility rather than for one token.

Tailwind generates all four utilities from those keys, so nothing here is a spelling mistake.

### Expected

Theme keys of every namespace classify the utilities Tailwind generates from them: `--text-*` to font size, `--shadow-*` to effects, `--spacing-*` to spacing, `--radius-*` to shape, and so on for the rest of the namespaces. The door for this looks like it already exists: `cn/config` exports `createCn` and `extendTailwindMerge`, both of which take `{ extend: { classGroups } }`, so feeding the parsed theme's non-colour keys into the config the plugin builds per project would cover all of these at once rather than one namespace at a time.

### Relation to open issues

- #8 is the `--text-*` half of this, reported as a colour. This issue is the same cause across the other namespaces, and the `no-restyle` side of it, so it is fine to fold into #8 if you would rather track one.
- #7 is the mirror case inside the colour namespace: theme keys Tailwind accepts that the declared-token set does not read.
- #4 is the sibling for `@utility` classes, which are also generated by the project and also unrecognized.

All four read as one missing step: the project's own theme is parsed for suggestions but never feeds classification.

### Versions

`@shadcn/lint` 0.1.0, Oxlint 1.83.0, Tailwind CSS v4.

**Comment by Leafgard** [comment]
Same cause, one more data point from a real Tailwind 4 + shadcn project, and the workaround we settled on.

Our theme declares a 13-step type scale as `--text-*` tokens (`--text-micro`, `--text-eyebrow`, `--text-value`, ...). On first run `no-raw-colors` reported 102 findings. 97 were these font-size utilities flagged as undeclared colours. The other five were literal `fill="#..."` on a brand logo. On a project with its own type scale the rule is about 95% noise until this lands.

The `--spacing-*` half shows up through `no-restyle` too, as the comment above predicts. `--spacing-control-h` and `--spacing-chip-h` generate `min-h-control-h` and `h-chip-h`. Passed to `<Textarea>` or `<Skeleton>` they are reported as `the grammar does not recognize it` whatever `allow` contains, so the only way out is naming each class in a contract.

What we run in the meantime, with both rules kept on:

```js
// eslint.config.mjs
const TYPOGRAPHY_UTILITIES = ['text-micro', 'text-eyebrow', /* one per --text-* token */]

rules: {
  'shadcn/no-raw-colors': ['error', { allow: TYPOGRAPHY_UTILITIES }],
  'shadcn/no-restyle': ['error', {
    allow: ['layout'],
    contracts: [{ pattern: '^Skeleton$', allow: ['layout', 'h-chip-h', 'h-control-h'] }],
  }],
}
```

Reading the `--text-*` and `--spacing-*` namespaces (and `--shadow-*`, per #10) from the resolved theme before classification would remove both lists. I can test a fix against this project.

**Comment by CarterOgunsola** [comment]
Same thing on a design system with a full type scale in `@theme` (`--text-body-small`, `--text-headline-3`, ...): every `text-<role>` use is reported as an undeclared color while `no-unknown-classes` stays quiet, so Tailwind knows the class and only the color rule is wrong. `@utility` names with a color prefix hit the same path (`bg-card-scrim` in our case, and the shadow ones in #10).

Cause is `classifierFor` filing unknown `text-*`/`bg-*`/`shadow-*` values under the color group, and `no-raw-colors` never consulting the `--text-*` declarations or `knownClassesFor().utilities` that `project/theme.ts` already reads. A lookup against those two sets before `undeclaredVerdict` would cover both this and #10.

Workaround we ship in our policy meanwhile: `"shadcn/no-raw-colors": ["error", { allow: ["text-body-*", "text-headline-*", "bg-card-scrim"] }]`. It works but hides typos of those names.
